### PR TITLE
Ruby-CarolynQi

### PR DESCRIPTION
**Pull request by qicarolyn1**
Submitting Adagrams

### CODE REVIEW

File: tests/test_wave_01.py [hunk 1]
```diff
@@ -31,13 +31,15 @@
     'Z': 1
 }
 
+#@pytest.mark.skip(reason="no way of currently testing this")
```
Comment (mmcknett-ada) [inline tests/test_wave_01.py:34]: [nit] You can remove these

File: adagrams/game.py [hunk 2]
```diff
@@ -1,11 +1,104 @@
...
+    final_letter_choice = []
+    for number in range(0,10):
+        chosen_letter = random.choice(list(letter_pool.keys()))
+        if (letter_pool[chosen_letter] == 1):
+            letter_pool.pop(chosen_letter)
+        else:
+            letter_pool[chosen_letter] -= 1
+        final_letter_choice.append(chosen_letter)
+        
+    return final_letter_choice
```
Comment (mmcknett-ada) [inline adagrams/game.py:41]: This way of choosing letters randomly doesn't quite meet the requirements. Notice that your solution is just as likely to draw a `Z` as it is to draw an `E`. The requirements say, "Since there are 12 `E`s but only 1 `Z`, it should be 12 times as likely for the user to draw an `E` as a `Z`". It turns out, our tests don’t test for that. (That test is quite hard to write!) Think about how you might modify your code to fulfill this requirement.

Regarding `random.choice`: when bringing in functions from imported modules (that we haven’t covered in the curriculum), please consider including a comment about how you understand the function to work, and how you researched it. Now that we’ve discussed Big O, it’s also useful to think about what the time/space complexity of the function might be. Thinking about how we can implement this functionality ourselves can help us gain insight into what that might be.

Remember that not every language has the same built-in functions, so we should be able to implement this same functionality ourselves. In the projects, we don’t ask for any features that we don’t think you should be able to write yourself. For drawing a random hand, the only external function that would be “required” is a way to pick a random number (such as randint). At this stage in our coding journeys, it’s often more valuable to re-implement things by hand to practice thinking about how to break down and approach the problem than to use a single library function to solve the problem for us.

File: adagrams/game.py [hunk 3]
```diff
@@ -1,11 +1,104 @@
...
+
 
 def uses_available_letters(word, letter_bank):
-    pass
+    letter_bank_new = letter_bank.copy()
+    for characters in word:
```
Comment (mmcknett-ada) [inline adagrams/game.py:46]: [nit] I usually like my loop variables to be singular because it gets assigned each member of a collection

```python
    for character in word:
```

File: adagrams/game.py [hunk 4]
```diff
@@ -1,11 +1,104 @@
...
+        try:
+            letter_bank_new.remove(characters.upper())
+        except ValueError:
+            return False
```
Comment (mmcknett-ada) [inline adagrams/game.py:50]: I wanted to point out how "Pythonic" this is. Python developers prefer to "ask for forgiveness" as you've done here. Great job! Also be aware that if you do this in any other language, especially Java, JavaScript, or C++, other devs are likely to take issue with it. Devs who write those languages prefer the "exceptions should be exceptional" approach. It's a way the Python community is unique. 😄

File: adagrams/game.py [hunk 5]
```diff
@@ -1,11 +1,104 @@
...
+        'V': 4, 
+        'W': 4, 
+        'X': 8, 
+        'Y': 4, 
+        'Z': 10
+    }
+    final_score = 0
+    for characters in word:
+        final_score += letter_score[characters.upper()]
+    if len(word) in range(7, 11):
```
Comment (mmcknett-ada) [inline adagrams/game.py:85]: [nit] `if len(word) >= 7:` might be less unexpected

File: adagrams/game.py [hunk 6]
```diff
@@ -1,11 +1,104 @@
...
+        if this_word_score > top_word_score:
+            winning_word = word
+            top_word_score = this_word_score
+        elif this_word_score == top_word_score:
+            if len(word) == 10 and len(winning_word) != 10:
+                winning_word = word
+                top_word_score = this_word_score
+            elif len(word) < len(winning_word) and len(winning_word) != 10:
+                winning_word = word
+                top_word_score = this_word_score
```
Comment (mmcknett-ada) [inline adagrams/game.py:103]: [suggestion] If you wanted to DRY up this `elif`, you could. Notice that both inner conditionals check `len(winning_word) != 10`. You could first refactor to combine that with the outer conditional like this:

```python
        elif this_word_score == top_word_score and len(winning_word) != 10:
            if len(word) == 10:
                winning_word = word
                top_word_score = this_word_score
            elif len(word) < len(winning_word):
                winning_word = word
                top_word_score = this_word_score
```

Then you could combine the two inner conditionals with an `or` like this:

```python
        elif this_word_score == top_word_score and len(winning_word) != 10:
            if len(word) == 10 or len(word) < len(winning_word):
                winning_word = word
                top_word_score = this_word_score
```

Comment (mmcknett-ada) [reply adagrams/game.py:103]: Note that you would *not* want to go further and DRY the outer `if/elif` with this same technique, because it just becomes too hard to read, even though it technically works correctly:

```python
        if this_word_score > top_word_score or this_word_score == top_word_score and len(winning_word) != 10 and (len(word) == 10 or len(word) < len(winning_word)):
            winning_word = word
            top_word_score = this_word_score
```

File: adagrams/game.py (hunk 5 above)
Comment (mmcknett-ada) [reply adagrams/game.py:85]: I noticed you changed away from this suggestion in one of your commits. The commit message only says "Another change", so I can't tell what the reasoning was behind the change. A more descriptive commit message might have prevented me from leaving this code review comment, which is excellent motivation for having more descriptive commit messages. 😉

**Review by mmcknett-ada — COMMENTED** [review]
Looks great! 🟢 Green!

I have some nitpicks, comments, and suggestions, but those don't detract from the fact that this is a solid submission. Very nice!

Some overall feedback:

- The tests are passing, and your code even passes a corner case I noticed during grading that aren't in the tests.
- I recommend committing as you finish sub-parts of waves. It can be easy to mess up previously-implemented functions as you work on new behavior, so it's valuable to have your history saved in git.
- There is one missing requirement (the `E` 12x as likely as `Z` one), but the tests can't actually test for it. I recommend revisiting and thinking about how you might implement that, as well. (Most people miss this requirement, including me the first time I did the problem.)
